src/org/intellij/erlang/psi/impl/ErlangFileImpl.java

```diff
@@ -712,6 +712,32 @@ public boolean process(PsiElement psiElement) {
 
   @NotNull
   @Override
+  public List<ErlangCallbackFunction> getOptionalCallbacks() {
+    ErlangFileStub stub = getStub();
+    if (stub != null) {
+      return getChildrenByType(stub, ErlangTypes.ERL_CALLBACK_FUNCTION, new ArrayFactory<ErlangCallbackFunction>() {
+        @NotNull
+        @Override
+        public ErlangCallbackFunction[] create(int count) {
+          return new ErlangCallbackFunction[count];
+        }
+      });
+    }
+
+    final List<ErlangCallbackFunction> result = ContainerUtil.newArrayList();
+    for (ErlangAttribute attribute : getAttributes()) {
+      ErlangOptionalCallbacks callbacks = attribute.getOptionalCallbacks();
+      ErlangOptionalCallbackFunctions functions = callbacks != null ? callbacks.getOptionalCallbackFunctions() : null;
+      if (functions == null) continue;
+      for (ErlangCallbackFunction function : functions.getCallbackFunctionList()) {
+        result.add(function);
+      }
+    }
+    return result;
```

Comment:
You compute these callback functions on every ```getOptionalCallbacks``` call. Take a look at other methods in this class - they all make use of caching - you should, too.